sycl/doc/extensions/proposed/sycl_ext_oneapi_barrier.asciidoc

```diff
@@ -0,0 +1,458 @@
+scope associated with a `sycl::atomic_ref`, and is orthogonal to the memory
+space in which the barrier object itself is allocated. The `Scope` template
+is effectively a promise that the barrier will only be used to synchronize
+threads of execution within the same scope, which may allow an implementation
+to improve performance. If the scope associated with a barrier is
+insufficiently broad to contain all threads of execution that attempt to use
+the barrier, behavior is undefined.
+
+[NOTE]
+====
+The threads of execution that are synchronized by this barrier may be a
+combination of host threads and device threads (i.e. work-items), if the
+barrier is visible to all threads of execution and has a sufficiently broad
+scope.
+====
+
+The completion function may be executed on any thread during a call to
+`arrive`, `arrive_and_drop` or `wait`. The completion function is considered
+device code, even when the barrier is used to synchronize a mixture of host and
+device threads, so it must abide by the SYCL restrictions for device code.
+
+This extension specification provides a brief overview of the functionality of
+the `barrier` class. If there are any differences in the definitions of this
+functionality, the wording in the {cpp}20 specification takes precedence.
+
+[source,c++]
+----
+namespace sycl::ext::oneapi::experimental {
+
+  template <sycl::memory_scope Scope, class CompletionFunction = __unspecified__>
+  class barrier {
+  public:
+    using arrival_token = __unspecified__;
+
+    static constexpr ptrdiff_t max() noexcept;
+
+    constexpr explicit barrier(ptrdiff_t expected,
+                               CompletionFunction f = CompletionFunction());
+    ~barrier();
+
+    barrier(const barrier&) = delete;
+    barrier& operator=(const barrier&) = delete;
+
+    [[nodiscard]] arrival_token arrive(ptrdiff_t update = 1);
+    void wait(arrival_token&& arrival) const;
+
+    void arrive_and_wait();
+    void arrive_and_drop();
+  };
+
+} // namespace sycl::ext::oneapi::experimental
+----
+
+[source,c++]
+----
+static constexpr ptrdiff_t max() noexcept;
+----
+_Returns_: The maximum number of threads of execution that can be synchronized
+by any `barrier` with the specified `Scope` and `CompletionFunction`.
+
+[source,c++]
+----
+constexpr explicit barrier(ptrdiff_t expected, CompletionFunction f = CompletionFunction());
+----
+_Preconditions_: If `Scope` is `memory_scope::work_group`, the calling thread
+of execution must be a work-item belonging to the work-group that will use the
+barrier. If `Scope` is `memory_scope::device`, the calling thread of execution
+must be executing on the device where the barrier will be used. If `Scope` is
+`memory_scope::system`, the calling thread of execution can be any host or
+device thread.
+
+_Effects_: Initializes the barrier with the given expected count and completion
+function, and begins the first phase of the barrier.
+
```

Comment:
Good point.  Added in 26f67e3931806b2f5ca0180481410326281c51f5.  This seems like a good place to clarify that the destructor cannot be called concurrently (which comes from `std::barrier`); I've added it so that it's clear we're not removing that restriction.